statistics/cmsketch.go

```diff
@@ -44,29 +57,210 @@ func NewCMSketch(d, w int32) *CMSketch {
+func (c *CMSketch) BuildTopN(data [][]byte, top, topNThreshold uint32, total uint64) {
+	sampleSize := uint64(len(data))
+	counter := groupElements(data)
+	sorted := make([]uint64, 0)
+
+	for _, vals := range counter {
+		for i := range vals {
+			sorted = append(sorted, vals[i].count)
+		}
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	ndv := uint32(len(sorted))
+	if top > ndv {
+		top = ndv
+	}
+
+	NthValue := sorted[top-1]
+	newNthValue := sorted[top-1]
+	sumTopN := uint64(0)
+
+	// Add a few elements, the number of which is close to the n-th most element.
+	for i := uint32(0); i < ndv && i < top*2; i++ {
+		// Here, 2/3 is get by running tests, tested 1, 1/2, 2/3, and 2/3 is relative better than 1 and 1/2
+		if i >= top && sorted[i]*3 < NthValue*2 && newNthValue != sorted[i] {
+			break
+		}
+		// sumTopN might be smaller than sum of final sum of elements in topNIndex.
+		// These two values are only used for build topNIndex, and they are not used in counting defaultValue.
+		newNthValue = sorted[i]
+		sumTopN += sorted[i]
+	}
+
+	estimateNDV, ratio, onlyOnceItems := calculateEstimateNDV(sorted, total)
+
+	enableTopN := (sampleSize/uint64(topNThreshold) <= sumTopN)
+	topN := make([]cmsCount, 0)
+	sumTopN = 0
+
+	for _, vals := range counter {
+		for i := range vals {
+			if enableTopN && vals[i].count >= newNthValue {
+				topN = append(topN, cmsCount{data: vals[i].data, count: vals[i].count * ratio})
+				sumTopN += vals[i].count * ratio
+			} else {
+				c.insertBytesN(vals[i].data, vals[i].count*ratio)
+			}
+		}
+	}
+	top = uint32(len(topN))
+	if enableTopN {
+		c.buildTopNMap(topN)
+	}
+
+	countWithoutTopN := total - (sampleSize-uint64(onlyOnceItems))*ratio
+
+	if total <= sumTopN {
+		c.defaultValue = 1
+	} else if estimateNDV <= uint64(top) {
+		c.defaultValue = 1
+	} else {
+		if estimateNDV+onlyOnceItems <= uint64(ndv) {
+			c.defaultValue = 1
+		} else {
+			c.defaultValue = countWithoutTopN / (estimateNDV - uint64(ndv) + onlyOnceItems)
+		}
+	}
+}
+
+func (c *CMSketch) buildTopNMap(topn []cmsCount) {
+	c.topNIndex = make(map[uint64][]cmsCount)
+	for i := range topn {
+		if topn[i].data == nil {
+			continue
+		}
+		h1, h2 := murmur3.Sum128(topn[i].data)
+		vals, ok := c.topNIndex[h1]
+		if !ok {
+			vals = make([]cmsCount, 0)
+		}
+		vals = append(vals, cmsCount{h1, h2, topn[i].data, topn[i].count})
+		c.topNIndex[h1] = vals
+	}
+}
+
+// queryAddTopN TopN adds count to CMSketch.topNIndex if exists, and returns the count of such elements after insert
+// if such elements does not in topn elements, nothing will happen and false will be returned.
+func (c *CMSketch) queryAddTopN(h1, h2, count uint64, d []byte) (uint64, bool) {
+	if c.topNIndex == nil {
+		return 0, false
+	}
+	cnt, ok := c.topNIndex[h1]
+	if !ok {
+		return 0, false
+	}
+	for k := range cnt {
+		if cnt[k].h2 == h2 && bytes.Equal(d, cnt[k].data) {
+			cnt[k].count += count
```

Comment:
I am afraid that `add 0` will also cause data race and lead to unexpected behavior.